ledger/src/blockstore_db.rs

```diff
@@ -1014,7 +1168,53 @@ impl<'a> WriteBatch<'a> {
     }
 }
 
-fn get_cf_options(access_type: &AccessType) -> Options {
+struct PurgedSlotFilter<C: Column + ColumnName>(Slot, CString, PhantomData<C>);
+
+impl<C: Column + ColumnName> CompactionFilter for PurgedSlotFilter<C> {
+    fn filter(&mut self, _level: u32, key: &[u8], _value: &[u8]) -> CompactionDecision {
+        use rocksdb::CompactionDecision::*;
+
+        let slot_in_key = C::slot(C::index(key));
+        if slot_in_key >= self.0 {
+            Keep
+        } else {
+            Remove
+        }
+    }
+
+    fn name(&self) -> &CStr {
+        &self.1
+    }
+}
+
+struct PurgedSlotFilterFactory<C: Column + ColumnName>(OldestSlot, CString, PhantomData<C>);
+
+impl<C: Column + ColumnName> CompactionFilterFactory for PurgedSlotFilterFactory<C> {
+    type Filter = PurgedSlotFilter<C>;
+
+    fn create(&mut self, _context: CompactionFilterContext) -> Self::Filter {
+        let copied_oldest_slot = self.0.get();
```

Comment:
hmmm is it possible for a range of keys to go a long time without compaction (especially those in the lower levels)? And if so, is the following a potential issue:

Imagining a scenario like:

1) We initialize the `OldestSlot` to default()

 2) Range `[A, B]` is compacted over and written to lowest level of SST's in the tree. No memory is cleaned because `if slot_in_key >= self.0` is true for all slots when `OldestSlot.get() == 0`.

3) We call `run_purge_with_stats()` from cleanup service over range `[A, C]` and set `OldestSlot` to ` B`, where `A < B < C`

Compaction over range `[A, C]` doesn't occur again and the memory isn't reclaimed.

Reply:
@carllin

> is it possible for a range of keys to go a long time without compaction (especially those in the lower levels)?

Virtually all column families (except `transaction_status_index`) are repeatedly compacted periodically (currently once per day) This is result of setting compaction_filter and lower `periodic_compaction_seconds`. And this should be applied to all sst files, regardless what levels they're at. This is also applies to the normal cfs (without primary_index) and special cfs (with primary_index).

Anyway, happy to see you're trying hard to find some potential issue. :)

> 1. We initialize the OldestSlot to default()

Yeah, I originally thought about serializing/deserializing `OldestSlot` but found no worth to do it. Let periodic `LedgerCleanupService` do it for us. And I think it's enough.

> 2. Range [A, B] is compacted over and written to lowest level of SST's in the tree. No memory is cleaned because if slot_in_key >= self.0 is true for all slots when OldestSlot.get() == 0.

Yeah, this is true. but these keys are re-compacted within 24 hours.

> We call run_purge_with_stats() from cleanup service over range [A, C] and set OldestSlot to B, where A < B < C

You mean set OldestSlot to C? That's what I intended to implement... each LedgerCleanupService purging iteration always sets `OldestSlot` to  the end of the purged range (`C` or `C+1` to be rigid).

> Compaction over range [A, C] doesn't occur again and the memory isn't reclaimed.

I'm seeing compactions are happening periodically for all sst files in test nodes. As compaction filter logic is opaque to RocksDb, RocksDb has no clue and just  applies compactions periodically.. ;) Then, eventually, the memory needed for keys will be reclaimed at that compaction.

Reply:
Ah ok I didn't know if the periodic compactions would reach the deepest levels of the LSM tree if those deepest levels weren't experiencing the memory pressure described herehttps://www.bookstack.cn/read/rocksdb-6.14-en/65581e1366cd2eb8.md 

`Compaction’s goal will be to restrict data size of those levels to be under the target`. 

I guess we write enough data where we trigger these compactions even on the deepest levels of the tree 😅, or we clean the old data before it reaches the very deep levels of the tree.

Reply:
@carllin 

>  I didn't know if the periodic compactions would reach the deepest levels of the LSM tree

Well, periodic compaction (my v2 and latest fix) doesn't reach the deepest levels by default. TTL compaction (my v1 fix) did. Also, not reaching the deepest levels is important for us not to cause huge key range compaction caused by seemingly-innocent modest delete_range by LedgerCleanupService for normal CFs. Also see more details for this at: https://github.com/solana-labs/solana/pull/16697#issuecomment-825662256


https://github.com/facebook/rocksdb/wiki/Leveled-Compaction#ttl:

> This also has the (good) side-effect of all the data in the non-bottommost level being newer than ttl, and all data in the bottommost level older than ttl. Note that it could lead to more writes as RocksDB would schedule more compactions.

So, ttl makes all sst files reach the deepest levels eventually.

Also, see my previous attempt to explain this: https://github.com/solana-labs/solana/pull/16697#issuecomment-826048374

https://github.com/facebook/rocksdb/wiki/Leveled-Compaction#periodic-compaction:

>  files ... will be eligible for compaction and be compacted **to the same level**.


Regardless of which levels sst files are, periodic compaction are run daily (for our case) without going down one deeper level, staying at the same level. Then, eventually all keys will be removed after being expired by our custom compaction filter.

Reply:
ah thanks for the explanation the "same level" part of `Whenever RocksDB tries to pick a compaction, files older than 30 days will be eligible for compaction and be compacted to the same level` is quite subtle.